ddtrace/tracer/spancontext.go

```diff
@@ -400,17 +401,20 @@ func (t *trace) finishedOne(s *span) {
 			s.setMeta(keyTraceID128, s.context.traceID.UpperHex())
 		}
 	}
-	if len(t.spans) != t.finished {
-		return
-	}
 	defer func() {
-		t.spans = nil
-		t.finished = 0 // important, because a buffer can be used for several flushes
+		if len(t.spans) == t.finished {
+			t.spans = nil
+			t.finished = 0 // important, because a buffer can be used for several flushes
+		}
 	}()
 	tr, ok := internal.GetGlobalTracer().(*tracer)
 	if !ok {
 		return
 	}
+	setPeerService(s, tr.config)
+	if len(t.spans) != t.finished {
+		return
+	}
```

Comment:
Because of this change, I start to think that `finishedOne` is not the right place to set peer service. Have you explored adding it in `StartSpan` instead? You have access to tracer config and span tags there + it's run on every span not only root spans.

Reply:
Hmm, sounds interesting, but what about the tags added after `StartSpan` when using `span.SetTag` after the span is started?

Reply:
You're right, it'd require to add a hook in `span.SetTag` as well to update the peer service accordingly.
Okay, it might be easier to keep it here in `finishedOne` but it requires some refactoring, the way it is implemented right now is brittle and not easy to read. I'm referring mainly to 
```
		if len(t.spans) == t.finished {
			t.spans = nil
			t.finished = 0 // important, because a buffer can be used for several flushes
		}
```
and
```
	if len(t.spans) != t.finished {
		return
	}
```
I'm still thinking we can improve this as I'm writing this comment.. 🤔

Reply:
It's not ideal, but removing the defer entirely and putting just `t.spans = nil` and `t.finished = 0` in the end of the function can be a good start.

Reply:
Would moving it back to the same order help?
```
        tr, ok := internal.GetGlobalTracer().(*tracer)
	if !ok {
		return
	}
	setPeerService(s, tr.config)
	if len(t.spans) != t.finished {
		return
	}
	defer func() {
		t.spans = nil
		t.finished = 0
	}()
```

Reply:
Ah, didn't load the previous comment

Reply:
> Would moving it back to the same order help?
> 
> ```
>         tr, ok := internal.GetGlobalTracer().(*tracer)
> 	if !ok {
> 		return
> 	}
> 	setPeerService(s, tr.config)
> 	if len(t.spans) != t.finished {
> 		return
> 	}
> 	defer func() {
> 		t.spans = nil
> 		t.finished = 0
> 	}()
> ```

This would be a subtle change in behavior because if the call to `GetGlobalTracer` fails, `return` is called prior to the `defer` being set.

Reply:
Is this a blocker for getting this PR merged, or can it be a followup improvement, as it is largely unrelated to the functionality that this PR is providing? There is a high level of urgency to get this merged and deployed to internal services for dogfooding.

Reply:
@ajgajg1134 and I are refactoring some of this code anyway as part of the partial flushing work, so we can rework some of that then. At the very least, I can confirm that this code doesn't change any behavior in the finished span other than:
1) we need to read an atomic value every span (which we'll have to do for partial flushing anyway, so that's fine for now)
2) peer service tags are set on every span in the trace upon finish

So I don't think we need to block on this